json_annotation/lib/src/allowed_keys_helpers.dart

```diff
@@ -2,16 +2,38 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-/// Helper function used in generated code when
+/// **DEPRECATED** Helper function used in generated code when
 /// `JsonSerializable.disallowUnrecognizedKeys` is `true`.
 ///
 /// Should not be used directly.
+@Deprecated('Code generated with the latest `json_serializable` will use '
+    '`\$checkKeys` instead. This function will be removed in the next majoy '
+    'release.')
 void $checkAllowedKeys(Map map, Iterable<String> allowedKeys) {
+  $checkKeys(map, allowedKeys: allowedKeys);
+}
+
+/// Helper function used in generated code when
+/// `JsonSerializable.disallowUnrecognizedKeys` is `true`.
+///
+/// Should not be used directly.
+void $checkKeys(Map map,
+    {Iterable<String> allowedKeys, List<String> requiredKeys}) {
   if (map == null) return;
-  var invalidKeys = map.keys.where((k) => !allowedKeys.contains(k));
-  if (invalidKeys.isNotEmpty) {
-    throw new UnrecognizedKeysException(
-        new List<String>.from(invalidKeys), map, allowedKeys.toList());
+
+  if (requiredKeys != null) {
+    var missingKeys = requiredKeys.where((k) => !map.keys.contains(k)).toList();
+    if (missingKeys.isNotEmpty) {
+      throw new MissingRequiredKeysException(missingKeys, map);
+    }
+  }
+
+  if (allowedKeys != null) {
+    var invalidKeys = map.keys.where((k) => !allowedKeys.contains(k));
+    if (invalidKeys.isNotEmpty) {
+      throw new UnrecognizedKeysException(
+          new List<String>.from(invalidKeys), map, allowedKeys.toList());
```

Comment:
lets move this `toList` up to the original assignment so its consistent

Reply:
ack